<!DOCTYPE html>

<html xmlns="http://www.w3.org/1999/xhtml">

<head>

<meta charset="utf-8" />
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />

<meta name="viewport" content="width=device-width, initial-scale=1">

<meta name="author" content="Yoonsu Cho" />


<title>Guide to using MR-TRYX</title>



<style type="text/css">code{white-space: pre;}</style>
<style type="text/css" data-origin="pandoc">
code.sourceCode > span { display: inline-block; line-height: 1.25; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode { white-space: pre; position: relative; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
code.sourceCode { white-space: pre-wrap; }
code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */

</style>
<script>
// apply pandoc div.sourceCode style to pre.sourceCode instead
(function() {
  var sheets = document.styleSheets;
  for (var i = 0; i < sheets.length; i++) {
    if (sheets[i].ownerNode.dataset["origin"] !== "pandoc") continue;
    try { var rules = sheets[i].cssRules; } catch (e) { continue; }
    for (var j = 0; j < rules.length; j++) {
      var rule = rules[j];
      // check if there is a div.sourceCode rule
      if (rule.type !== rule.STYLE_RULE || rule.selectorText !== "div.sourceCode") continue;
      var style = rule.style.cssText;
      // check if color or background-color is set
      if (rule.style.color === '' && rule.style.backgroundColor === '') continue;
      // replace div.sourceCode by a pre.sourceCode rule
      sheets[i].deleteRule(j);
      sheets[i].insertRule('pre.sourceCode{' + style + '}', j);
    }
  }
})();
</script>



<style type="text/css">body {
background-color: #fff;
margin: 1em auto;
max-width: 700px;
overflow: visible;
padding-left: 2em;
padding-right: 2em;
font-family: "Open Sans", "Helvetica Neue", Helvetica, Arial, sans-serif;
font-size: 14px;
line-height: 1.35;
}
#header {
text-align: center;
}
#TOC {
clear: both;
margin: 0 0 10px 10px;
padding: 4px;
width: 400px;
border: 1px solid #CCCCCC;
border-radius: 5px;
background-color: #f6f6f6;
font-size: 13px;
line-height: 1.3;
}
#TOC .toctitle {
font-weight: bold;
font-size: 15px;
margin-left: 5px;
}
#TOC ul {
padding-left: 40px;
margin-left: -1.5em;
margin-top: 5px;
margin-bottom: 5px;
}
#TOC ul ul {
margin-left: -2em;
}
#TOC li {
line-height: 16px;
}
table {
margin: 1em auto;
border-width: 1px;
border-color: #DDDDDD;
border-style: outset;
border-collapse: collapse;
}
table th {
border-width: 2px;
padding: 5px;
border-style: inset;
}
table td {
border-width: 1px;
border-style: inset;
line-height: 18px;
padding: 5px 5px;
}
table, table th, table td {
border-left-style: none;
border-right-style: none;
}
table thead, table tr.even {
background-color: #f7f7f7;
}
p {
margin: 0.5em 0;
}
blockquote {
background-color: #f6f6f6;
padding: 0.25em 0.75em;
}
hr {
border-style: solid;
border: none;
border-top: 1px solid #777;
margin: 28px 0;
}
dl {
margin-left: 0;
}
dl dd {
margin-bottom: 13px;
margin-left: 13px;
}
dl dt {
font-weight: bold;
}
ul {
margin-top: 0;
}
ul li {
list-style: circle outside;
}
ul ul {
margin-bottom: 0;
}
pre, code {
background-color: #f7f7f7;
border-radius: 3px;
color: #333;
white-space: pre-wrap; 
}
pre {
border-radius: 3px;
margin: 5px 0px 10px 0px;
padding: 10px;
}
pre:not([class]) {
background-color: #f7f7f7;
}
code {
font-family: Consolas, Monaco, 'Courier New', monospace;
font-size: 85%;
}
p > code, li > code {
padding: 2px 0px;
}
div.figure {
text-align: center;
}
img {
background-color: #FFFFFF;
padding: 2px;
border: 1px solid #DDDDDD;
border-radius: 3px;
border: 1px solid #CCCCCC;
margin: 0 5px;
}
h1 {
margin-top: 0;
font-size: 35px;
line-height: 40px;
}
h2 {
border-bottom: 4px solid #f7f7f7;
padding-top: 10px;
padding-bottom: 2px;
font-size: 145%;
}
h3 {
border-bottom: 2px solid #f7f7f7;
padding-top: 10px;
font-size: 120%;
}
h4 {
border-bottom: 1px solid #f7f7f7;
margin-left: 8px;
font-size: 105%;
}
h5, h6 {
border-bottom: 1px solid #ccc;
font-size: 105%;
}
a {
color: #0033dd;
text-decoration: none;
}
a:hover {
color: #6666ff; }
a:visited {
color: #800080; }
a:visited:hover {
color: #BB00BB; }
a[href^="http:"] {
text-decoration: underline; }
a[href^="https:"] {
text-decoration: underline; }

code > span.kw { color: #555; font-weight: bold; } 
code > span.dt { color: #902000; } 
code > span.dv { color: #40a070; } 
code > span.bn { color: #d14; } 
code > span.fl { color: #d14; } 
code > span.ch { color: #d14; } 
code > span.st { color: #d14; } 
code > span.co { color: #888888; font-style: italic; } 
code > span.ot { color: #007020; } 
code > span.al { color: #ff0000; font-weight: bold; } 
code > span.fu { color: #900; font-weight: bold; }  code > span.er { color: #a61717; background-color: #e3d2d2; } 
</style>




</head>

<body>




<h1 class="title toc-ignore">Guide to using MR-TRYX</h1>
<h4 class="author">Yoonsu Cho</h4>


<div id="TOC">
<ul>
<li><a href="#data-setup">Data setup</a></li>
<li><a href="#running-tryx">Running Tryx</a></li>
<li><a href="#step-by-step-analysis">Step-by-step analysis</a></li>
<li><a href="#significant-candidate-traits">Significant candidate traits</a></li>
<li><a href="#adjustment">Adjustment</a></li>
<li><a href="#visualisation">Visualisation</a></li>
</ul>
</div>

<p>The following analyses should run within a couple of minutes, depending on internet speed and the traffic that the <a href="https://gwas.mrcieu.ac.uk/">IEU GWAS database</a> servers are experiencing.</p>
<p>Begin by choosing an exposure-outcome hypothesis to explore. e.g. LDL cholesterol on coronary heart disease. These data can be extracted from the IEU GWAS database using the <a href="https://mrcieu.github.io/TwoSampleMR/">TwoSampleMR</a> package:</p>
<div id="data-setup" class="section level2">
<h2>Data setup</h2>
<p>If necessary install TwoSampleMR:</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1"></a>devtools<span class="op">::</span><span class="kw">install_github</span>(<span class="st">&quot;mrcieu/TwoSampleMR@ieugwasr&quot;</span>)</span></code></pre></div>
<p>Create a dataset for LDL and CHD</p>
<div class="sourceCode" id="cb2"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1"></a><span class="kw">library</span>(TwoSampleMR)</span>
<span id="cb2-2"><a href="#cb2-2"></a>a &lt;-<span class="st"> </span><span class="kw">extract_instruments</span>(<span class="st">&quot;ieu-a-300&quot;</span>)</span>
<span id="cb2-3"><a href="#cb2-3"></a><span class="co">#&gt; API: public: https://gwas-api.mrcieu.ac.uk/</span></span>
<span id="cb2-4"><a href="#cb2-4"></a><span class="co">#&gt; Warning in format_data(d, type = &quot;exposure&quot;, snps = NULL, phenotype_col =</span></span>
<span id="cb2-5"><a href="#cb2-5"></a><span class="co">#&gt; &quot;phenotype&quot;, : eaf column is not numeric. Coercing...</span></span>
<span id="cb2-6"><a href="#cb2-6"></a>b &lt;-<span class="st"> </span><span class="kw">extract_outcome_data</span>(a<span class="op">$</span>SNP, <span class="st">&quot;ieu-a-7&quot;</span>, <span class="dt">access_token=</span><span class="ot">NULL</span>)</span>
<span id="cb2-7"><a href="#cb2-7"></a><span class="co">#&gt; Extracting data for 81 SNP(s) from 1 GWAS(s)</span></span>
<span id="cb2-8"><a href="#cb2-8"></a><span class="co">#&gt; Finding proxies for 1 SNPs in outcome ieu-a-7</span></span>
<span id="cb2-9"><a href="#cb2-9"></a><span class="co">#&gt; Extracting data for 1 SNP(s) from 1 GWAS(s)</span></span>
<span id="cb2-10"><a href="#cb2-10"></a>dat &lt;-<span class="st"> </span><span class="kw">harmonise_data</span>(a,b)</span>
<span id="cb2-11"><a href="#cb2-11"></a><span class="co">#&gt; Harmonising LDL cholesterol || id:ieu-a-300 (ieu-a-300) and Coronary heart disease || id:ieu-a-7 (ieu-a-7)</span></span>
<span id="cb2-12"><a href="#cb2-12"></a><span class="co">#&gt; Removing the following SNPs for incompatible alleles:</span></span>
<span id="cb2-13"><a href="#cb2-13"></a><span class="co">#&gt; rs10947332</span></span>
<span id="cb2-14"><a href="#cb2-14"></a><span class="co">#&gt; Removing the following SNPs for being palindromic with intermediate allele frequencies:</span></span>
<span id="cb2-15"><a href="#cb2-15"></a><span class="co">#&gt; rs2954029, rs3184504, rs964184</span></span></code></pre></div>
</div>
<div id="running-tryx" class="section level2">
<h2>Running Tryx</h2>
<p>First load the package</p>
<p>We can now perform the analysis:</p>
<div class="sourceCode" id="cb3"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1"></a><span class="kw">library</span>(tryx)</span>
<span id="cb3-2"><a href="#cb3-2"></a><span class="co">#&gt; Loading required package: dplyr</span></span>
<span id="cb3-3"><a href="#cb3-3"></a><span class="co">#&gt; </span></span>
<span id="cb3-4"><a href="#cb3-4"></a><span class="co">#&gt; Attaching package: &#39;dplyr&#39;</span></span>
<span id="cb3-5"><a href="#cb3-5"></a><span class="co">#&gt; The following objects are masked from &#39;package:stats&#39;:</span></span>
<span id="cb3-6"><a href="#cb3-6"></a><span class="co">#&gt; </span></span>
<span id="cb3-7"><a href="#cb3-7"></a><span class="co">#&gt;     filter, lag</span></span>
<span id="cb3-8"><a href="#cb3-8"></a><span class="co">#&gt; The following objects are masked from &#39;package:base&#39;:</span></span>
<span id="cb3-9"><a href="#cb3-9"></a><span class="co">#&gt; </span></span>
<span id="cb3-10"><a href="#cb3-10"></a><span class="co">#&gt;     intersect, setdiff, setequal, union</span></span>
<span id="cb3-11"><a href="#cb3-11"></a><span class="co">#&gt; Loading required package: RadialMR</span></span>
<span id="cb3-12"><a href="#cb3-12"></a><span class="co">#&gt; Loading required package: magrittr</span></span>
<span id="cb3-13"><a href="#cb3-13"></a><span class="co">#&gt; Loading required package: tidyr</span></span>
<span id="cb3-14"><a href="#cb3-14"></a><span class="co">#&gt; </span></span>
<span id="cb3-15"><a href="#cb3-15"></a><span class="co">#&gt; Attaching package: &#39;tidyr&#39;</span></span>
<span id="cb3-16"><a href="#cb3-16"></a><span class="co">#&gt; The following object is masked from &#39;package:magrittr&#39;:</span></span>
<span id="cb3-17"><a href="#cb3-17"></a><span class="co">#&gt; </span></span>
<span id="cb3-18"><a href="#cb3-18"></a><span class="co">#&gt;     extract</span></span>
<span id="cb3-19"><a href="#cb3-19"></a><span class="co">#&gt; Loading required package: ggplot2</span></span>
<span id="cb3-20"><a href="#cb3-20"></a><span class="co">#&gt; Loading required package: glmnet</span></span>
<span id="cb3-21"><a href="#cb3-21"></a><span class="co">#&gt; Loading required package: Matrix</span></span>
<span id="cb3-22"><a href="#cb3-22"></a><span class="co">#&gt; </span></span>
<span id="cb3-23"><a href="#cb3-23"></a><span class="co">#&gt; Attaching package: &#39;Matrix&#39;</span></span>
<span id="cb3-24"><a href="#cb3-24"></a><span class="co">#&gt; The following objects are masked from &#39;package:tidyr&#39;:</span></span>
<span id="cb3-25"><a href="#cb3-25"></a><span class="co">#&gt; </span></span>
<span id="cb3-26"><a href="#cb3-26"></a><span class="co">#&gt;     expand, pack, unpack</span></span>
<span id="cb3-27"><a href="#cb3-27"></a><span class="co">#&gt; Loaded glmnet 3.0-1</span></span>
<span id="cb3-28"><a href="#cb3-28"></a><span class="co">#&gt; Loading required package: ggrepel</span></span>
<span id="cb3-29"><a href="#cb3-29"></a>x &lt;-<span class="st"> </span>Tryx<span class="op">$</span><span class="kw">new</span>(dat)</span>
<span id="cb3-30"><a href="#cb3-30"></a>x<span class="op">$</span><span class="kw">mrtryx</span>()</span>
<span id="cb3-31"><a href="#cb3-31"></a><span class="co">#&gt; Using RadialMR package to detect outliers</span></span>
<span id="cb3-32"><a href="#cb3-32"></a><span class="co">#&gt; </span></span>
<span id="cb3-33"><a href="#cb3-33"></a><span class="co">#&gt; Radial IVW</span></span>
<span id="cb3-34"><a href="#cb3-34"></a><span class="co">#&gt; </span></span>
<span id="cb3-35"><a href="#cb3-35"></a><span class="co">#&gt;                   Estimate  Std.Error   t value     Pr(&gt;|t|)</span></span>
<span id="cb3-36"><a href="#cb3-36"></a><span class="co">#&gt; Effect (Mod.2nd) 0.4131547 0.04848067  8.522050 1.567537e-17</span></span>
<span id="cb3-37"><a href="#cb3-37"></a><span class="co">#&gt; Iterative        0.4131605 0.04848234  8.521877 1.569885e-17</span></span>
<span id="cb3-38"><a href="#cb3-38"></a><span class="co">#&gt; Exact (FE)       0.4307847 0.02360690 18.248257 2.136159e-74</span></span>
<span id="cb3-39"><a href="#cb3-39"></a><span class="co">#&gt; Exact (RE)       0.4182297 0.05616141  7.446922 1.251321e-10</span></span>
<span id="cb3-40"><a href="#cb3-40"></a><span class="co">#&gt; </span></span>
<span id="cb3-41"><a href="#cb3-41"></a><span class="co">#&gt; </span></span>
<span id="cb3-42"><a href="#cb3-42"></a><span class="co">#&gt; Residual standard error: 2.056 on 76 degrees of freedom</span></span>
<span id="cb3-43"><a href="#cb3-43"></a><span class="co">#&gt; </span></span>
<span id="cb3-44"><a href="#cb3-44"></a><span class="co">#&gt; F-statistic: 72.63 on 1 and 76 DF, p-value: 1.1e-12</span></span>
<span id="cb3-45"><a href="#cb3-45"></a><span class="co">#&gt; Q-Statistic for heterogeneity: 321.246 on 76 DF , p-value: 6.778753e-32</span></span>
<span id="cb3-46"><a href="#cb3-46"></a><span class="co">#&gt; </span></span>
<span id="cb3-47"><a href="#cb3-47"></a><span class="co">#&gt;  Outliers detected </span></span>
<span id="cb3-48"><a href="#cb3-48"></a><span class="co">#&gt; Number of iterations = 3</span></span>
<span id="cb3-49"><a href="#cb3-49"></a><span class="co">#&gt; Warning in if (radial$outliers[1] == &quot;No significant outliers&quot;) {: the</span></span>
<span id="cb3-50"><a href="#cb3-50"></a><span class="co">#&gt; condition has length &gt; 1 and only the first element will be used</span></span>
<span id="cb3-51"><a href="#cb3-51"></a><span class="co">#&gt; Identified 7 outliers</span></span>
<span id="cb3-52"><a href="#cb3-52"></a><span class="co">#&gt; Using default list of 288 traits</span></span>
<span id="cb3-53"><a href="#cb3-53"></a><span class="co">#&gt; Extracting data for 7 SNP(s) from 288 GWAS(s)</span></span>
<span id="cb3-54"><a href="#cb3-54"></a><span class="co">#&gt; Found 24 candidate traits associated with outliers at p &lt; 5e-08</span></span>
<span id="cb3-55"><a href="#cb3-55"></a><span class="co">#&gt; Finding instruments for candidate traits</span></span>
<span id="cb3-56"><a href="#cb3-56"></a><span class="co">#&gt; Warning in format_data(d, type = &quot;exposure&quot;, snps = NULL, phenotype_col =</span></span>
<span id="cb3-57"><a href="#cb3-57"></a><span class="co">#&gt; &quot;phenotype&quot;, : eaf column is not numeric. Coercing...</span></span>
<span id="cb3-58"><a href="#cb3-58"></a><span class="co">#&gt; Removing outlier SNPs from candidate trait instrument lists</span></span>
<span id="cb3-59"><a href="#cb3-59"></a><span class="co">#&gt; 24 traits with at least one instrument</span></span>
<span id="cb3-60"><a href="#cb3-60"></a><span class="co">#&gt; Looking up candidate trait instruments for Coronary heart disease || id:ieu-a-7</span></span>
<span id="cb3-61"><a href="#cb3-61"></a><span class="co">#&gt; Extracting data for 730 SNP(s) from 1 GWAS(s)</span></span>
<span id="cb3-62"><a href="#cb3-62"></a><span class="co">#&gt; 710 instruments extracted for Coronary heart disease || id:ieu-a-7</span></span>
<span id="cb3-63"><a href="#cb3-63"></a><span class="co">#&gt; Looking up candidate trait instruments for LDL cholesterol || id:ieu-a-300</span></span>
<span id="cb3-64"><a href="#cb3-64"></a><span class="co">#&gt; Extracting data for 730 SNP(s) from 1 GWAS(s)</span></span>
<span id="cb3-65"><a href="#cb3-65"></a><span class="co">#&gt;  instruments extracted for LDL cholesterol || id:ieu-a-300</span></span>
<span id="cb3-66"><a href="#cb3-66"></a><span class="co">#&gt; Looking up exposure instruments for 24 candidate traits</span></span>
<span id="cb3-67"><a href="#cb3-67"></a><span class="co">#&gt; Extracting data for 77 SNP(s) from 24 GWAS(s)</span></span>
<span id="cb3-68"><a href="#cb3-68"></a><span class="co">#&gt; 1790 instruments extracted</span></span>
<span id="cb3-69"><a href="#cb3-69"></a><span class="co">#&gt; Removing outlier SNPs from candidate trait outcome lists</span></span>
<span id="cb3-70"><a href="#cb3-70"></a><span class="co">#&gt; Removed 46 outlier SNPs</span></span>
<span id="cb3-71"><a href="#cb3-71"></a><span class="co">#&gt; Performing MR of 24 candidate traits against Coronary heart disease || id:ieu-a-7</span></span>
<span id="cb3-72"><a href="#cb3-72"></a><span class="co">#&gt; Performing MR of 23 candidate traits against LDL cholesterol || id:ieu-a-300</span></span>
<span id="cb3-73"><a href="#cb3-73"></a><span class="co">#&gt; Performing MR of LDL cholesterol || id:ieu-a-300 against 24 candidate traits</span></span></code></pre></div>
<p>This will do the following:</p>
<ol style="list-style-type: decimal">
<li>Find outlier SNPs in the exposure-outcome analysis</li>
<li>Find traits in the IEU GWAS database that those outliers associate with. These traits are known as ‘candidate traits’</li>
<li>Extract instruments for those ‘candidate traits’</li>
<li>Perform MR of each of those ‘candidate traits’ against the exposure and the outcome</li>
</ol>
</div>
<div id="step-by-step-analysis" class="section level2">
<h2>Step-by-step analysis</h2>
<p>The analysis steps can be done one-by-one running the following commands:</p>
<div class="sourceCode" id="cb4"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb4-1"><a href="#cb4-1"></a><span class="co"># Find outlier SNPs in the exposure-outcome analysis</span></span>
<span id="cb4-2"><a href="#cb4-2"></a>x<span class="op">$</span><span class="kw">get_outliers</span>()</span>
<span id="cb4-3"><a href="#cb4-3"></a></span>
<span id="cb4-4"><a href="#cb4-4"></a><span class="co"># Find traits in the MR-Base database that those outliers associate with. These traits are known as &#39;candidate traits&#39;</span></span>
<span id="cb4-5"><a href="#cb4-5"></a>x<span class="op">$</span><span class="kw">set_candidate_traits</span>()</span>
<span id="cb4-6"><a href="#cb4-6"></a>x<span class="op">$</span><span class="kw">scan</span>()</span></code></pre></div>
<p>Note that you can browse available traits here: <a href="https://gwas.mrcieu.ac.uk/" class="uri">https://gwas.mrcieu.ac.uk/</a> and get a complete list using:</p>
<div class="sourceCode" id="cb5"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1"></a>traits &lt;-<span class="st"> </span>TwoSampleMR<span class="op">::</span><span class="kw">available_outcomes</span>()</span></code></pre></div>
<div class="sourceCode" id="cb6"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb6-1"><a href="#cb6-1"></a><span class="co"># Extract instruments for those &#39;candidate traits&#39;</span></span>
<span id="cb6-2"><a href="#cb6-2"></a>x<span class="op">$</span><span class="kw">extractions</span>() <span class="co">#which includes the following functions:</span></span>
<span id="cb6-3"><a href="#cb6-3"></a>  x<span class="op">$</span><span class="kw">candidate_instruments</span>()</span>
<span id="cb6-4"><a href="#cb6-4"></a>  x<span class="op">$</span><span class="kw">outcome_instruments</span>()</span>
<span id="cb6-5"><a href="#cb6-5"></a>  x<span class="op">$</span><span class="kw">exposure_instruments</span>()</span>
<span id="cb6-6"><a href="#cb6-6"></a>  x<span class="op">$</span><span class="kw">exposure_candidate_instruments</span>()</span>
<span id="cb6-7"><a href="#cb6-7"></a></span>
<span id="cb6-8"><a href="#cb6-8"></a><span class="co"># Make datasets for MR analysis</span></span>
<span id="cb6-9"><a href="#cb6-9"></a>x<span class="op">$</span><span class="kw">harmonise</span>() <span class="co">#which includes the following functions:</span></span>
<span id="cb6-10"><a href="#cb6-10"></a>  x<span class="op">$</span><span class="kw">candidate_outcome_dat</span>()</span>
<span id="cb6-11"><a href="#cb6-11"></a>  x<span class="op">$</span><span class="kw">candidate_exposure_dat</span>()</span>
<span id="cb6-12"><a href="#cb6-12"></a>  x<span class="op">$</span><span class="kw">exposure_candidate_dat</span>()</span>
<span id="cb6-13"><a href="#cb6-13"></a></span>
<span id="cb6-14"><a href="#cb6-14"></a><span class="co"># Perform MR of each of those &#39;candidate traits&#39; against the exposure and the outcomes</span></span>
<span id="cb6-15"><a href="#cb6-15"></a>x<span class="op">$</span><span class="kw">mr</span>()</span></code></pre></div>
<p>See the <code>?Tryx</code> for options on the parameters for this analysis. e.g. You can specify your own set of outliers, for example SNPs that have extreme p-values in the outcome GWAS</p>
<div class="sourceCode" id="cb7"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb7-1"><a href="#cb7-1"></a>a &lt;-<span class="st"> </span><span class="kw">as.character</span>(<span class="kw">subset</span>(dat, pval.outcome <span class="op">&lt;</span><span class="st"> </span><span class="fl">5e-8</span>)<span class="op">$</span>SNP)</span>
<span id="cb7-2"><a href="#cb7-2"></a>x &lt;-<span class="st"> </span>Tryx<span class="op">$</span><span class="kw">new</span>(dat)</span>
<span id="cb7-3"><a href="#cb7-3"></a>x<span class="op">$</span><span class="kw">mrtryx</span>(<span class="dt">outliers=</span>a)</span></code></pre></div>
<p>The next steps are to determine which of the candidate traits are of interest (e.g. using p-value thresholds), visualise the results, and adjust the exposure-outcome estimates based on knowledge of the ‘candidate trait’ associations.</p>
</div>
<div id="significant-candidate-traits" class="section level2">
<h2>Significant candidate traits</h2>
<p>One can determine which of the putative associations might be ‘interesting’ in different ways. We have provided a simple convenience function to apply different multiple testing corrections. e.g.</p>
<div class="sourceCode" id="cb8"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb8-1"><a href="#cb8-1"></a>x<span class="op">$</span><span class="kw">tryx.sig</span>()</span></code></pre></div>
<p>Will by default use FDR of 5%. See <code>?tryx.sig</code> for more options.</p>
</div>
<div id="adjustment" class="section level2">
<h2>Adjustment</h2>
<p>Finally, to adjust the SNP effects on the exposure and outcome traits given their influences on the candidate traits, we can run:</p>
<div class="sourceCode" id="cb9"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb9-1"><a href="#cb9-1"></a>x<span class="op">$</span><span class="kw">adjustment</span>()</span>
<span id="cb9-2"><a href="#cb9-2"></a>x<span class="op">$</span><span class="kw">adjustment.mv</span>()</span>
<span id="cb9-3"><a href="#cb9-3"></a>x<span class="op">$</span><span class="kw">analyse</span>()</span>
<span id="cb9-4"><a href="#cb9-4"></a>x<span class="op">$</span><span class="kw">analyse.mv</span>()</span></code></pre></div>
<p>By default, this adjusts for the trait that has the largest impact for a particular SNP.</p>
<p>The <code>x$output$adjustment$estimates</code> show the adjusted effect estimates. A plot is generated showing how SNP effects have changed due to candidate trait adjustments in <code>x$output$adjustment$plot</code>.</p>
</div>
<div id="visualisation" class="section level2">
<h2>Visualisation</h2>
<p>To produce a basic diagram of the connectivity of SNPs, candidate traits, exposure and outcome:</p>
<div class="sourceCode" id="cb10"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb10-1"><a href="#cb10-1"></a><span class="kw">tryx.network</span>(tryxscan)</span></code></pre></div>
<p>This shows that some candidate traits influence the exposure only, the outcome only, or both the exposure and the outcome.</p>
<p>You can also create a volcano plot of the candidate-exposure and/or candidate-outcome associations. e.g. to show exposures and outcomes</p>
<div class="sourceCode" id="cb11"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb11-1"><a href="#cb11-1"></a><span class="kw">volcano_plot</span>(</span>
<span id="cb11-2"><a href="#cb11-2"></a>    <span class="kw">rbind</span>(tryxscan<span class="op">$</span>candidate_exposure_mr, tryxscan<span class="op">$</span>candidate_outcome_mr)</span>
<span id="cb11-3"><a href="#cb11-3"></a>)</span></code></pre></div>
<p>or e.g. just exposures</p>
<div class="sourceCode" id="cb12"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb12-1"><a href="#cb12-1"></a><span class="kw">volcano_plot</span>(tryxscan<span class="op">$</span>candidate_exposure_mr)</span></code></pre></div>
</div>



<!-- code folding -->


<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
